Become plugins

bcoca opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Go from hardcoded become 'methods' to plugins so users can add their own or override the built-in as needed

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
become

acozine commented:
@bcoca you can fix this crop of `not-in-toctree` errors by adding `docs/docsite/rst/plugins/become.rst` to match the other plugin-type pages, and including a "glob" TOC:
```
.. toctree:: :maxdepth: 1
    :glob:

    become/*
```

bcoca commented:
@acozine I was planning on writing that file last, but i'll put in placeholder to avoid the errors

jctanner commented:
bot_status

jborean93 commented on lib/ansible/plugins/become/runas.py:
```diff
@@ -0,0 +1,74 @@
+              - name: ANSIBLE_BECOME_FLAGS
+              - name: ANSIBLE_RUNAS_FLAGS
+    notes:
+        - runas is really implemented in the winrm conneciton plugin and as such can only be used with winrm connections.
```
It's more implemented in the powershell shell plugin than winrm. We have both psrp and winrm that support runas as they are both Windows specific. When finally adding support for SSH connections to Windows then it would also be supported there.

quater committed:
[WIP] become plugins

Move from hardcoded method to plugins for ease of use, expansion and overrides
  - load into connection as it is going to be the main consumer
  - play_context will also use to keep backwards compat API
  - ensure shell is used to construct commands when needed
  - migrate settings remove from base config in favor of plugin specific configs
  - cleanup ansible-doc
  - add become plugin docs
  - remove deprecated sudo/su code and keywords
  - adjust become options for cli
  - set plugin options from context
  - ensure config defs are avaialbe before instance
  - refactored getting the shell plugin, fixed tests
     - changed into regex as they were string matching, which does not work with random string generation
     - explicitly set flags for play context tests
 - moved plugin loading up front
 - now loads for basedir also
 - allow pyc/o for non m modules
 - fixes to tests and some plugins
 - migrate to play objects fro play_context
 - simiplify gathering
 -  added utf8 headers
 - moved option setting
 - add fail msg to dzdo
 - use tuple for multiple options on fail/missing
 - fix relative plugin paths
 - shift from play context to play
 - all tasks already inherit this from play directly
 - remove obsolete 'set play'
 - correct environment handling
 - add wrap_exe option to pfexec
 - fix runas to noop
 - fixed setting play context
 - added password configs
 - removed required false
 - remove from doc building till they are ready

future development:
  - deal with 'enable' and 'runas' which are not 'command wrappers' but 'state flags' and currently hardcoded in diff subsystems

jamescassell commented on lib/ansible/plugins/become/dzdo.py:
```diff
@@ -0,0 +1,86 @@
+              - name: ANSIBLE_DZDO_EXE
+        become_flags:
+            description: Options to pass to dzdo
+            default: -H -S -n
```
Where is this default set?

sivel replied:
Right where you commented.  The `ConfigManger` uses the `DOCUMENTATION` from plugins to configure those plugins.

jamescassell replied:
Awesome!

sivel commented on lib/ansible/cli/playbook.py:
```diff
@@ -89,27 +90,26 @@ def run(self):
                 raise AnsibleError("the playbook: %s could not be found" % playbook)
             if not (os.path.isfile(playbook) or stat.S_ISFIFO(os.stat(playbook).st_mode)):
                 raise AnsibleError("the playbook: %s does not appear to be a file" % playbook)
+            pb_dirs.append(os.path.dirname(playbook))
```
This line is the reason that `The module helloworld_linux was not found in configured module paths` is happening.

`os.path.dirname(playbook)` on something like `test_binary_modules.yml` just becomes `''`.

```suggestion
            pb_dirs.append(os.path.dirname(os.path.abspath(playbook)))
```

Otherwise, you can handle using `abspath` in `add_all_plugin_dirs`

sivel commented on lib/ansible/plugins/connection/__init__.py:
```diff
@@ -83,30 +83,30 @@ def __init__(self, play_context, new_stdin, shell=None, *args, **kwargs):
-            raise AnsibleError("Invalid shell type specified (%s), or the plugin for that shell type is missing." % shell_type)
+            self._shell = get_shell_plugin(shell_type=getattr(self, '_shell_type', None), executable=self._play_context.executable)
+
+        self._become = None
```
Since we are using this outside of the connection (e.g. accessing `getattr(self._connection, '_become')`), we should remove the `_` prefix that indicates it is private.

bcoca replied:
i would like to remedy that by moving most of those into connection/shell and out of action plugins, which should not care this much about these details.

sivel commented:
I haven't looked into why exactly, but the `TypeError: 'NoneType' object is not iterable` coming from `ShellBase.chown` is caused by `self._play_context.become_user` being `None` when `_fixup_perms2` is called from `ActionBase._execute_module`.

Because of this `_is_become_unprivileged` is returning `True`, when it shouldn't be.

If I am reading this correctly, I think that we need to use `self._connection._become.get_option('become_user')`

However we have places that expected to be setting `self._play_context.become_user`, so those places may need to update the plugin options.

jborean93 commented on lib/ansible/plugins/become/runas.py:
```diff
@@ -0,0 +1,74 @@
+              - name: ANSIBLE_BECOME_USER
+              - name: ANSIBLE_RUNAS_USER
+            required: True
+        become_exe:
```
Does this need to be set? We don't call an executable at all with runas so this would do nothing.

bcoca replied:
it doesn't, i kept mostly for 'parity' with other plugins

jborean93 commented on lib/ansible/plugins/become/runas.py:
```diff
@@ -0,0 +1,74 @@
+              - name: ANSIBLE_RUNAS_EXE
+        become_flags:
+            description: Options to pass to runas
+            default: -H -S -n
```
By default we don't set any flags.

bcoca replied:
fxied

jborean93 commented on lib/ansible/plugins/become/runas.py:
```diff
@@ -0,0 +1,74 @@
+              - name: ANSIBLE_RUNAS_FLAGS
+        become_pass:
+            description: password
+            required: False
```
I assume `required: False` is the default, would it need to be explicitly set?

jborean93 commented on lib/ansible/plugins/become/runas.py:
```diff
@@ -0,0 +1,74 @@
+            description: password
+            required: False
+    notes:
+        - runas is really implemented in the powershell module handler and as such can only be used with winrm connections.
```
Probably change `winrm` to `Windows` as `psrp`, and soon `ssh`, will support runas against Windows

jborean93 commented on lib/ansible/config/base.yml:
```diff
@@ -1059,50 +1068,6 @@ DEFAULT_SU:
-  default: False
-  deprecated:
-    why: In favor of Ansible Become, which is a generic framework
-    version: "2.9"
```
It says this will be removed in 2.9 not 2.8 so sounds like this is being removed a version too early.

jborean93 replied:
Ahh I see they have been added to the sudo plugin, ignore me here.

jborean93 commented on lib/ansible/plugins/become/runas.py:
```diff
@@ -0,0 +1,74 @@
+            env:
+              - name: ANSIBLE_BECOME_EXE
+              - name: ANSIBLE_RUNAS_EXE
+        become_flags:
```
While not necessary right now it would be nice if become_flags could take in a dict, or convert key=value, to a dict. Right now become_flags in runas is based on specific key=values and having the controller do the validation vs the remote execution would be nice.

alikins commented on lib/ansible/config/base.yml:
```diff
@@ -329,12 +329,20 @@ ACTION_WARNINGS:
   version_added: "2.5"
+BECOME_PLUGIN_PATH:
+  name: Become plugins path
+  default: ~/.ansible/plugins/become:/usr/share/ansible/become/action
```
`/usr/share/ansible/become/action` ?  should this be /usr/share/ansible/plugins/become ?

bcoca replied:
good catch, bad edit after copy+paste

alikins commented on lib/ansible/config/base.yml:
```diff
@@ -329,12 +329,20 @@ ACTION_WARNINGS:
+BECOME_PLUGIN_PATH:
+  name: Become plugins path
+  default: ~/.ansible/plugins/become:/usr/share/ansible/become/action
+  description: Colon separated paths in which Ansible will search for Action Plugins.
```
s/Action/Become

bcoca replied:
he, same copy paste issue

alikins commented on lib/ansible/config/base.yml:
```diff
@@ -329,12 +329,20 @@ ACTION_WARNINGS:
   - {key: action_warnings, section: defaults}
   type: boolean
   version_added: "2.5"
+BECOME_PLUGIN_PATH:
```
Most of the plugin path config items are of form 'DEFAULT_CACHE_PLUGIN_PATH' etc. 

If there isn't a particular reason for BECOME_PLUGIN_PATH instead of DEFAULT_BECOME_PLUGIN_PATH, may be better to keep it consistent with naming of other plugin path config items.

bcoca replied:
i'm actually  moving away from DEFAULT since that is a misleading tag

abadger commented on docs/docsite/Makefile:
```diff
@@ -26,7 +26,7 @@ ifdef PLUGINS
 endif
 
-DOC_PLUGINS ?= cache callback connection inventory lookup shell strategy vars
+DOC_PLUGINS ?= become cache callback cliconf connection httpapi inventory lookup shell strategy vars
```
Shouldn't cliconf and httpapi be added in a different PR?  It looks like the landing pages for those two don't currently exist.

bcoca replied:
removed from list but added comment

abadger commented on lib/ansible/cli/__init__.py:
```diff
@@ -339,29 +340,7 @@ def ask_passwords(self):
-            _dep('sudo')
-        if self.options.ask_su_pass or self.options.su_user:
-            _dep('su')
+        pass
```
This isn't part of our stable API so we should remove the calls to it and remove the method itself.

bcoca replied:
i'm fine with removing it, i was just trying to be 'nice'

alikins commented on lib/ansible/executor/task_executor.py:
```diff
@@ -832,12 +831,18 @@ def _get_connection(self, variables, templar):
                         variables[i] = delegated_vars[i]
 
-        conn_type = self._play_context.connection
+        # load become plugin if needed
```
Would it be worthwhile to add a default of no-op become plugin to always have a plugin to avoid the if/elses?
(Could possibly make things like plugins.connection.ConnectionBase.load_become simpler or possibly unneeded)

bcoca replied:
I had been thinking about that, a noop plugin is easy enough, the current 'runas' is 99% that.

abadger commented on lib/ansible/cli/__init__.py:
```diff
@@ -496,32 +466,18 @@ def base_parser(usage="", output_opts=False, runas_opts=False, meta_opts=False,
-                                   (C.DEFAULT_BECOME_METHOD, ' | '.join(C.BECOME_METHODS)))
+            runas_group.add_option('--become-method', dest='become_method', default=C.DEFAULT_BECOME_METHOD,
+                                   help="privilege escalation method to use (default=%s), use `ansible-doc -t become -l` to list valid choices."
+                                   % (C.DEFAULT_BECOME_METHOD))
```
We just changed something else to list out the valid methods.  We may want to do the same here.

@gundalow

bcoca replied:
the other change (list of documentable plugin types) still relied on constants, which won't be the case with plugins

abadger commented on lib/ansible/playbook/block.py:
```diff
@@ -370,7 +370,7 @@ def _get_parent_attribute(self, attr, extend=False, prepend=False):
     def filter_tagged_tasks(self, play_context, all_vars):
```
remove play_context from the method's signature since it's no longer used.

bcoca replied:
i kept for backwards compatibility, again its not needed and it is not really 'public api' but wanted to minimize breakage for others.

alikins commented on lib/ansible/playbook/__init__.py:
```diff
@@ -40,17 +39,18 @@
 class Playbook:
 
-    def __init__(self, loader):
+    def __init__(self, loader, options=None):
```
Passing a command line options instance (a optparse.Values ) deep down into Playbook and Play objects seems wrong to me. It couples ansible.playbook.* stuff to a particular cli arg parser (things like ansible.playbook.FieldAttributeBase._set_option have to deal with the implementation details of optparse)

bcoca replied:
possibly, this way i saw as 'least intrusive' since now we do it by 'side loading' into play_context, this keeps the correct precedence (cli sets default for keywords).

bcoca replied:
we can change it slightly and make 'options' a dict which is populated from the cli object, but I didn't think we needed that at this point

abadger replied:
@alikins +1

jimi-c replied:
We've talked about converting options to a dict for 3+ years... it would be nice to finally do it.

bcoca replied:
I don't disagree, but i would not make it a requirement here

alikins commented on lib/ansible/playbook/play_context.py:
```diff
@@ -446,130 +380,30 @@ def set_task_and_variable_override(self, task, variables, templar):
-        prompt = None
-        success_key = None
-        self.prompt = None
+        if cmd and self.become:
```
I'd flip this and return early and deindent the rest of the method

(completely untested example)

``` python
   def make_become_cmd(self, cmd, executable=None):
        if not cmd or not self.become:
            return cmd

        # load/call become plugins here
        plugin = become_loader.get(self.become_method)
        if not plugin:
            raise AnsibleError("Privilege escalation method not found: %s" % self.become_method)

        options = {'become_exe': self.become_exe or getattr(self, '%s_exe' % self.become_method, self.become_method) or self.become_method,
                   'become_flags': self.become_flags or getattr(self, '%s_flags' % self.become_method, '') or '',
                   'become_user': self.become_user,
                   'become_pass': self.become_pass}
        plugin.set_options(direct=options)

        if not executable:
            executable = self.executable

        shell = get_shell_plugin(executable=executable)
        cmd = plugin.build_become_command(cmd, shell)

        # for backwards compat:
        self.prompt = plugin.prompt

        return cmd
```

bcoca replied:
i prefer to keep 1 return, it seems to lose a lot of people when you have more than one exit from the function

abadger replied:
@alikins +1  That is better coding style in this case.   If we have a short bit of logic that removes certain specific cases from the rest of the code, it makes sense to return early.  That way we don't have to wonder throughout the rest of the function what happens to the other cases.  https://youtu.be/rrBJVMyD-Gs?t=755

bcoca committed:
cleanup

  remove callers to removed func
  removed --sudo cli doc refs
  remove runas become_exe
  ensure keyerorr on plugin
  also fix backwards compat, missing method is attributeerror, not ansible error
  get remote_user consistently
  ignore missing system_tmpdirs on plugin load
  correct config precedence
  add deprecation
  fix networking imports
  backwards compat for plugins using BECOME_METHODS

abadger commented on lib/ansible/playbook/base.py:
```diff
@@ -231,12 +233,37 @@ def load_data(self, ds, variable_manager=None, loader=None):
+        # set defaults from CLI
+        if options:
+            self._set_options(options)
+
```
This looks very very fragile.  In addition to what alikins said earlier, we're taking an arbitrary data structure and overwriting attributes in the dictionary if they exist.  If they don't exist, then we're discarding them.  This removes the programming language's builtin abilities to tell when bad data is being passed to it.

This also doesn't seem like it is an inherent part of a become plugin so even if we want to try it, it should be part of a different PR.

bcoca replied:
its would be a prerequisite, since we need a way to get options into the task object from which become consumes.

abadger replied:
We should be sending specific options in or (better design) the object should be requesting the defaults that it needs.

bcoca replied:
i agree, but for this change i just copied what we do in play_context, I would leave redoing the API for a subsequent PR ... probably after we remove play_context from the picture.

abadger replied:
It probably should go into a PR before this one.

The easiest thing to do is probably to save the CLI options to a program global once it is parsed.  We currently use ansible.constants in that manner so ansible.constants.cli_options might be a suitable name.

abadger commented on lib/ansible/playbook/block.py:
```diff
@@ -367,10 +367,10 @@ def _get_parent_attribute(self, attr, extend=False, prepend=False):
         '''
         Creates a new block, with task lists filtered based on the tags contained
-        within the play_context object.
+        within the play object. play_context is not in use anymore
```
remove play_context from here too.

bcoca commented:
bot_status

bcoca commented:
bot_broken

jimi-c commented on lib/ansible/config/base.yml:
```diff
@@ -329,12 +329,20 @@ ACTION_WARNINGS:
   description:
     - By default Ansible will issue a warning when the shell or command module is used and the command appears to be similar to an existing Ansible module.
-    - These warnings can be silenced by adjusting this setting to False. You can also control this at the task level with the module option ``warn``.
+    - These warnings Can be silenced by adjusting this setting to False. You can also control this at the task level with the module option ``warn``.
```
Typo here.

bcoca replied:
you spoil my hidden emphasis! .. fixed

jimi-c commented on lib/ansible/playbook/play_context.py:
```diff
@@ -443,130 +377,30 @@ def set_task_and_variable_override(self, task, variables, templar):
 
-        if self.become:
+            # load/call become plugins here
+            plugin = become_loader.get(self.become_method)
```
I don't understand why we'd keep any of this here anymore, as it could all be done later when the become_loader is used in the connection plugin. With this, we're pretty much just double-loading the plugin (with caching maybe we don't care) but it's just very disjointed, and trying to find where this is done when all the other become stuff has moved out of PlayContext seems confusing.

bcoca replied:
its a backwards compatibility issue, i kept in case 3rd party code still calls this, none of our current code in this PR does so it won't double load for us.

bcoca replied:
it was also a good way to run existing unit tests

sivel commented on lib/ansible/constants.py:
```diff
@@ -68,32 +69,14 @@ def set_constant(name, value, export=vars()):
     export[name] = value
 
 
+def BECOME_METHODS():
```
I don't think switching this to a callable, will achieve the result that is expected.  People won't be calling it, so they just wouldn't get anything they expect.

bcoca replied:
no, the constant needs to populate from the callable, this is mostly a failed experiment

bcoca replied:
the only way to do this with the desired behaviour would be to inject a class as the constants module and have a getter for the specific value ... not sure that is somewhere i want to go.

bcoca committed:
better constant update

bcoca committed:
fix become_plugins for playbooks

abadger commented:
Being fixed here: https://github.com/ansible/ansible/pull/50421
